_posts/2018-09-06-stateless-microservices.md

```diff
@@ -0,0 +1,272 @@
+
+```bash
+$ curl http://127.0.0.1:8080/function/frank-says/
+Frank has entered the building.
+```
+
+You can also try a custom route:
+
+```
+$ curl http://127.0.0.1:8080/function/frank-says/logout
+Frank has left the building.
+```
+
+You can try updating the messages or adding some other routes then run `faas-cli up` again to redeploy the microservice.
+
+Now check out `faas-cli list` to see the invocation count rising each time you access the microservice.
+
+```
+$ faas-cli list
+Function                        Invocations     Replicas
+frank-says                      5               1
+```
+
+#### Trigger auto-scaling
+
+We can now trigger auto-scaling with a simple bash `for` loop:
+
+```
+for i in {1..10000} ; do sleep 0.01 && curl http://127.0.0.1:8080/function/frank-says && echo ; done
+```
+
+In another window enter: `watch faas-cli list` or run `faas-cli list` periodically. You should see the value for `Inovcations` increase and the `Replicas` increase as auto-scaling kicks in.
+
+```
+Function                        Invocations     Replicas
+frank-says                    	702            	4 
+```
+
+When your bash `for` loop completes or when you cancel it with Ctrl+C you will see the replica count decrease back to `1`.
+
+You can also use the OpenFaaS UI to monitor and invoke the microservice at http://127.0.0.1:8080
+
+![](/images/stateless-microservices/frank-says.png)
+
+Read more on auto-scaling including how to [configure min, max and zero replica scaling parameters](https://docs.openfaas.com/architecture/autoscaling/).
+
+### Deploy the Sinatra guestbook with MySQL
+
+```
+git clone https://github.com/openfaas-incubator/openfaas-sinatra-guestbook
+cd openfaas-sinatra-guestbook
+```
+
+Configure your MySQL database details in `./sql.yml`. If you don't have access to MySQL you can deploy it using [helm](https://github.com/helm/charts/tree/master/stable/mysql) on Kubernetes.
+
+```
+cp sql.example.yml sql.yml
+```
+
+Finally deploy the guestbook:
+
+```
+faas-cli up
+
+http://127.0.0.1:8080/function/guestbook
+```
+
+Use the URL given to you by the command above to access the microservice.
+
+Sign the guest book using the UI and when you're done you can reset the MySQL table at any time by posting to `/function/guestbook/reset`.
+
+![](/images/stateless-microservices/signed.png)
+
+This guestbook is stateless so it stores its state in a MySQL table meaning each auto-scaled replica gets the exact same view of the world. This is a key part of a stateless microservice.
+
+#### Enable Zero-Scale
+
+To enable zero-scale, simply [follow the documentation](https://docs.openfaas.com/architecture/autoscaling/) to enable the `faas-idler`.
+
+Then add a label to your `stack.yml` file to tell OpenFaaS that your function is eligible for zero-scaling:
+
+`com.openfaas.scale.zero: true`
+
+Finally redeploy the guestbook with `faas-cli up`. The faas-idler will now scale your function to zero replicas as soon as it is detected as idle.
```

Comment:
Maybe mention some details, like what's the timing and how it gets scaled up again.

Reply:
Done